server/auto_updater/dataset_updater.py

```diff
@@ -0,0 +1,305 @@
+            while True:
+                try:
+
+                    pmc_paper = open_stream.next()
+
+                    if pmc_paper is None:
+                        break
+
+                    if pmc_paper.isdir():
+                        continue
+
+                    if pmc_paper.name in current_pmc_set:
+                        continue
+
+                    new_paper = open_stream.extractfile(pmc_paper)
+                    doc_vector, word_counter = generate_vector_counts(
+                        word_model, new_paper, "//abstract/sec/*|//abstract/p|//body/sec/*|//body/p"
+                    )
+                    
+                    dir_writer.writerow(
+                        {"tarfile": tar_file, "file_path": str(pmc_paper.name)}
+                    )
+
+                    if word_counter is None:
+                        continue
+
+                    embed_writer.writerow(
+                        {
+                            "document": Path(pmc_paper.name).stem,
+                            "journal": str(Path(pmc_paper.name).parent),
+                            **dict(
+                                zip([f"feat_{idx}" for idx in range(300)], doc_vector)
+                            ),
+                        }
+                    )
+
+                    for tok in word_counter:
+                        count_writer.writerow(
+                            {
+                                "document": Path(pmc_paper.name).stem,
+                                "lemma": tok,
+                                "count": word_counter[tok],
+                            }
+                        )
+
+                except tarfile.ReadError as e:
+                    print(
+                        f"File {pmc_open_access_url}{tar_file} achieved an error: {str(e)}"
+                    )
+                    print("Skipping...")
+                    break
+
+            open_stream.close()
+            requested_file_stream.close()
+
+
+def generate_vector_counts(model, document_path, xpath, filter_tags=filter_tag_list):
+    """
+    This method is designed to construct document vectors for a given xml document.
+    Every document has specific tags that are striped in order to have accurate embeddings
+
+    Keyword arguments:
+        model - the model to extract word vectors from
+        xpath_str - the xpath string to extract tags from the xml document
+        filter_tag_list - the list of tags to strip from the xml document
+    """
+
+    word_vectors = []
+
+    tree = ET.parse(document_path, parser=parser)
+
+    # Process xml without specified tags
+    ET.strip_tags(tree, *filter_tags)
+
+    root = tree.getroot()
+    
+    if root.attrib['article-type'].strip() != 'research-article':
+        return [], None
+    
+    all_text = root.xpath(xpath)
+    all_text = list(map(lambda x: "".join(list(x.itertext())), all_text))
+    all_text = " ".join(all_text)
+
```

Comment:
@danich1 I was able to gather all new papers from the tarball files (in series, not in parallel). In order to boost the performance, I added the following code between lines 197-198:
```
    # Optimization: Remove stop words from `all_text` before feeding it to nlp.
    # This optimization not only speeds up the data processing 5%-10%, but also
    # minimizes memory usage.
    all_text = [x for x in all_text.split() if x not in nlp.Defaults.stop_words]
    all_text = " ".join(all_text)

    # Set nlp.max_length dynamically
    if nlp.max_length < len(all_text):
        nlp.max_length = len(all_text)
```
See the comments on why I made the change. Does it make sense to you?

The whole process took 5 hours and 14 minutes on my local machine. Will test it in parallel.

Reply:
@danich1: By the way, as far as I understand, it's okay to have duplicate papers in `new_papers_added_dir.tsv` ("file_path" column), but the paper IDs must be unique in `new_papers_embedding.tsv` ("document" column) and `new_paper_token_count.tsv` ("document" column), right?

Reply:
>See the comments on why I made the change. Does it make sense to you?

Yeah this makes sense. I would expect a big speed up once you do things in parallel.

>@danich1: By the way, as far as I understand, it's okay to have duplicate papers in new_papers_added_dir.tsv ("file_path" column), but the paper IDs must be unique in new_papers_embedding.tsv ("document" column) and new_paper_token_count.tsv ("document" column), right?

Correct.